### DISCUSSION TITLE
payment plugin: how to resume in-progress transaction?

### DISCUSSION THREAD

**Discussion by stapelberg** [pretix/pretix#3229] [Q&A] [answered]
I implemented a payment plugin which delegates the entire payment to a third-party provider via HTTP redirects. The flow is:

1. User starts payment
2. My plugin’s `execute_payment()` starts a transaction with the payment provider and redirects the user to the provider, with the success_url set to a return URL.
3. The user pays or cancels, the provider redirects to the return URL.
4. In the return view, my plugin fetches the transaction status from the payment provider and confirms the payment if successful, or redirects back to the order if unsuccessful.

---

This works fine in general, but we noticed that one user had trouble and ended up paying twice.

I think what happened is this:

1. User started payment
2. `execute_payment()` starts transaction 1 and redirects
3. user scans the QR code on their mobile device, **closes the browser tab**, but **also confirms the payment** on the mobile device.
4. the return URL is never accessed, pretix never learns about the new payment status.
5. the user is confused and tries again.
6. `execute_payment()` creates transaction 2 and redirects
7. User pays successfully (again!).

---

Now, I’m not sure what would be the best way to fix this issue. I thought that maybe I could persist the transaction id into the payment.info_data (that works!) and access it from `execute_payment()` — unfortunately that doesn’t work because pretix creates a new payment when the user retries. 

So, one possible fix would be for `change_payment_provider()` in `pretix/base/services/orders.py` to not cancel the old payment when the payment provider is the same provider as before.

Another fix might be within the plugin: we could query the payment provider for completed transactions for `payment.order.code`, query their status, and mark the payment as successful immediately. My concern with this approach is the potential for confusion when there are legitimate reasons to pay the same order twice, e.g. a pay, refund, pay situation.

Or maybe another fix would be in the UI/dispatch logic: instead of changing the payment provider, we could call a plugin function to try and confirm a payment if there is a pending one?

Or maybe I’m missing something else altogether? Thanks in advance for any tips

**Comment by pc-coholic** [comment, marked as answer]
Well - ideally the payment provider offers some kind of webhook - loosing the user on the return path happens quite a lot and everything else is quite a futile struggle to keep things going ;-)

Other things you could do:
- Check the payment status in the `payment_pending_render` and check if the payment has been already completed
- Have a `periodic_task` check the `canceled` and `pending` order payments if they have been completed. That way, at least you can catch the double payments and mark the orders accordingly in the backend for refund. But do ideally include some kind of filter on how far back you're going with the payment-checks - else you'll be soon checking payments from 3 years ago ;-) Perhaps the payment-provider has a natural expiration on each payment that you can leverage?

But yeah - when it boils down to it: Use Webhooks and/or pester you payment provider until they provide webhooks. This not 1995 ;-)

--- 

Edit
> Another fix might be within the plugin: we could query the payment provider for completed transactions for payment.order.code, query their status, and mark the payment as successful immediately. My concern with this approach is the potential for confusion when there are legitimate reasons to pay the same order twice, e.g. a pay, refund, pay situation.

If possible, do not use the Order Code as the UUID but rather the `payment.full_id` so that it expands to something like `ORDER-P-1` - those are unique per order and thus won't impede if you have more than one payment per order.

**Reply by stapelberg** [reply]
Thanks for the pointer!

The documentation of the payment provider is very sparse, but after looking very carefully, I have indeed found a webhook feature that should be triggered even if the user doesn’t hit the return url.

I’ll try to get this implemented, and also use the full_id as reference number.

Update: I successfully implemented webhook support, see https://github.com/stapelberg/pretix-datatrans/commit/312c51670cb985b6776192cae90a02459701bd40 in case anyone else would like an example.

**Reply by raphaelm** [reply]
Looks good! :)

**Reply by stapelberg** [reply]
Thanks for taking a look! Much appreciated
